### PR TITLE
adds rolling window support

### PR DESCRIPTION
**Pull request by rsiddharthan**
Adds rolling window feature to a tablecloth dataset. This initial work supports integer and time based indices, and simple fixed length windows which contain the set of rows. preceding the current row.

### CODE REVIEW

**Comment by rsiddharthan** [comment]
Refactored code and switched to using updated slice function.

TODO #1. Switch to the new group-by that supports ordered datasets
TODO #2. Discuss if rolling-window can operate on index. It will require changed to slice to do the same

**Comment by rsiddharthan** [comment]
Checked in version is an ordered dataset.

TODO: Add sample test cases for time based indices.

**Comment by rsiddharthan** [comment]
Checked in support for column-name in function signatures. 
This allows both int based and time based indices to work

**Comment by ezmiller** [comment]
@rsiddharthan the tests are failing, I think because midje needs to be included in the project dependencies in project.clj. The error:
>    Could not locate midje/sweet__init.class, midje/sweet.clj or midje/sweet.cljc
   on classpath.

**Comment by ezmiller** [comment]
@rsiddharthan I added the midje dependency to project.clj, but I'm still getting an error:
```clj
 Syntax error compiling at (tablecloth/time/api/rolling_window_test.clj:7:1).
Unable to find column {{:loc 0}=[0], {:loc 1}=[0 1], {:loc 2}=[0 1 2], {:loc 3}=[1 2 3], {:loc 4}=[2 3 4], {:loc 5}=[3 4 5], {:loc 6}=[4 5 6], {:loc 7}=[5 6 7], {:loc 8}=[6 7 8], {:loc 9}=[7 8 9]}
```
Tried to understand it but nothing jumped out at me. Maybe it will be familiar to you.

**Comment by ezmiller** [comment]
@rsiddharthan the command I ran to run your tests on your new test file on the command line was
```
lein midje tablecloth.time.api.rolling-window-test
```

**Review by ezmiller — CHANGES_REQUESTED** [review]
@rsiddharthan submitting my review now, now that we've go the tests worked out. I'll mark it "request changes" only because of the comment about the `import` statements that should be moved to the `ns` form.

File: src/tablecloth/time/api/rolling_window.clj [hunk 1]
```diff
@@ -0,0 +1,93 @@
+(ns tablecloth.time.api.rolling-window
+  (:require [tablecloth.api :as tablecloth]
```
[the remark this answers is not in the document]
Comment (rsiddharthan) [reply src/tablecloth/time/api/rolling_window.clj:3]: updated code to move the java imports

File: src/tablecloth/time/api/rolling_window.clj [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            [tablecloth.time.api :refer [slice]]))
+(import java.util.Map)
+(import java.util.LinkedHashMap)
```
[the remark this answers is not in the document]
Comment (rsiddharthan) [reply src/tablecloth/time/api/rolling_window.clj]: updated

File: src/tablecloth/time/api/rolling_window.clj [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  "creates an ordered map {k, v} from ordered inputs [k] [v])"
+  [k v]
+  (let [om (LinkedHashMap.)
+        kv (mapv vector k v)]
+    (doseq [[ke ve] kv]
+      (.put ^Map om ke ve))
+    om))
+
+
+(defn- instructions
```
[the remark this answers is not in the document]
Comment (rsiddharthan) [reply src/tablecloth/time/api/rolling_window.clj]: renamed to rw-map. This remains a private function, as I prefer the single public rolling-window function

**Review by ezmiller — APPROVED** [review]
@rsiddharthan the changes look good. I will approve, but I think we want to get to the bottom of the odd diffs related to the slice tests and the change to support in slice, lest we end up messing up `main` somehow. 

Also looks like there is a merge conflict which is not surprising in the least, so we'll need to merge main into this branch before we can merge. Happy to help with that process if it's unfamiliar.

**Pull request merged**